engine/src/witness/hub.rs

```diff
@@ -0,0 +1,367 @@
+								signer_id: epoch.info.0,
+								tx_fee,
+								tx_metadata: (),
+								transaction_ref: PolkadotTransactionId {
+									block_number: header.index,
+									extrinsic_index
+								}
+							}
+							.into(),
+							epoch.index,
+						)
+						.await;
+					}
+				},
+			Err(error) => {
+				// We expect this to occur when attempting to decode
+				// a transaction that was not sent by us.
+				// We can safely ignore it, but we log it in case.
+				tracing::debug!("Failed to decode UncheckedExtrinsic {error}");
+			},
+		}
+	}
+}
+
+pub async fn start<StateChainClient, ProcessCall, ProcessingFut>(
+	scope: &Scope<'_, anyhow::Error>,
+	dot_client: DotRetryRpcClient,
+	process_call: ProcessCall,
+	state_chain_client: Arc<StateChainClient>,
+	state_chain_stream: impl StreamApi<FINALIZED> + Clone,
+	epoch_source: EpochSourceBuilder<'_, '_, StateChainClient, (), ()>,
+	db: Arc<PersistentKeyDB>,
+) -> Result<()>
+where
+	StateChainClient: StorageApi + SignedExtrinsicApi + 'static + Send + Sync,
+	ProcessCall: Fn(state_chain_runtime::RuntimeCall, EpochIndex) -> ProcessingFut
+		+ Send
+		+ Sync
+		+ Clone
+		+ 'static,
+	ProcessingFut: Future<Output = ()> + Send + 'static,
+{
+	let unfinalised_source = HubUnfinalisedSource::new(dot_client.clone())
+		.strictly_monotonic()
+		.then(|header| async move { header.data.iter().filter_map(filter_map_events).collect() })
+		.shared(scope);
+
+	unfinalised_source
+		.clone()
+		.chunk_by_time(epoch_source.clone(), scope)
+		.chain_tracking(state_chain_client.clone(), dot_client.clone())
+		.logging("chain tracking")
+		.spawn(scope);
+
+	let epoch_source = epoch_source
+		.filter_map(
+			|state_chain_client, _epoch_index, hash, _info| async move {
+				state_chain_client
+					.storage_value::<pallet_cf_environment::AssethubVaultAccountId<state_chain_runtime::Runtime>>(
+						hash,
+					)
+					.await
+					.expect(STATE_CHAIN_CONNECTION)
+			},
+			|_state_chain_client, _epoch, _block_hash, historic_info| async move { historic_info },
+		)
+		.await;
+
+	let vaults = epoch_source.vaults::<cf_chains::Assethub>().await;
+
+	// Full witnessing
+	HubFinalisedSource::new(dot_client.clone())
+		.strictly_monotonic()
+		.logging("finalised block produced")
+		.then(|header| async move {
+			header.data.iter().filter_map(filter_map_events).collect::<Vec<_>>()
+		})
+		.chunk_by_vault(vaults, scope)
+		.deposit_addresses(scope, state_chain_stream.clone(), state_chain_client.clone())
+		.await
+		// Deposit witnessing
+		.hub_deposits(process_call.clone())
+		// Proxy added witnessing
+		.then(proxy_added_witnessing)
+		// Broadcast success
+		.egress_items(scope, state_chain_stream.clone(), state_chain_client.clone())
+		.await
+		.then({
+			let process_call = process_call.clone();
+			let dot_client = dot_client.clone();
+			move |epoch, header| {
+				process_egress(epoch, header, process_call.clone(), dot_client.clone())
+			}
+		})
+		.continuous("Polkadot".to_string(), db)
```

Comment:
this is a particularly important change, otherwise polkadot and assethub would be writing to the same database table
```suggestion
		.continuous("Assethub".to_string(), db)
```

Reply:
I see, thanks for catching that!